activemq-unit-tests/src/test/java/org/apache/activemq/jms2/ActiveMQJMS2ContextTest.java

```diff
@@ -296,24 +296,57 @@ public void testProducerDeliveryDelaySet() throws JMSException {
         messageProducer.setDeliveryDelay(1000l);
     }
 
-    @Test(expected = UnsupportedOperationException.class)
+    @Test
     public void testProducerSendMessageCompletionListener() throws JMSException {
-         messageProducer.send(session.createQueue(methodNameDestinationName), null, (CompletionListener)null);
+        messageProducer.send(session.createQueue(methodNameDestinationName), session.createTextMessage("test message"), null);
     }
 
-    @Test(expected = UnsupportedOperationException.class)
+    @Test
     public void testProducerSendMessageQoSParamsCompletionListener() throws JMSException {
-         messageProducer.send(null, 1, 4, 0l, null);
+         messageProducer.send(session.createTextMessage("test message"), 1, 4, 0l, null);
     }
 
-    @Test(expected = UnsupportedOperationException.class)
+    @Test
     public void testProducerSendDestinationMessageCompletionListener() throws JMSException {
-         messageProducer.send(session.createQueue(methodNameDestinationName), null, null);
+         messageProducer.send(session.createQueue(methodNameDestinationName), session.createTextMessage("test message"), null);
     }
 
-    @Test(expected = UnsupportedOperationException.class)
+    @Test
     public void testProducerSendDestinationMessageQosParamsCompletionListener() throws JMSException {
-         messageProducer.send(session.createQueue(methodNameDestinationName), null, 1, 4, 0l, null);
+         messageProducer.send(session.createQueue(methodNameDestinationName), session.createTextMessage("test message"), 1, 4, 0l, null);
+    }
+
+    @Test
+    public void testProducerSendMessageWithNonNullCompletionListener() throws JMSException {
+        final String testTextMessageBody = "this is a test message";
+        final CompletionListener completionListener = new CompletionListener() {
+            @Override
+            public void onCompletion(Message message) {
+                TextMessage textMessage = (TextMessage) message;
+                try {
+                    assertEquals(textMessage.getText(), testTextMessageBody);
+                } catch (JMSException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+
+            @Override
+            public void onException(Message message, Exception e) {
+                throw new RuntimeException(e);
+            }
+        };
+        messageProducer.send(
+                session.createQueue(methodNameDestinationName),
+                session.createTextMessage(testTextMessageBody),
+                completionListener);
```

Comment:
Its also more than a little odd to use ActiveMQJMS2ContextTest to test this, since you cant created a MessageProducer with a JMSContext, but rather only JMSProducer instances.

This is largely an existing issue though since all the other existing tests you updated, those checking for throwing UnsupportedOperationException, do the same. Seems like maybe those are c&p errors from long ago.